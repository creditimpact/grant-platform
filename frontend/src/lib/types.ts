export type CaseStatus =
  | 'empty'
  | 'open'
  | 'submitted'
  | 'processing'
  | 'complete'
  | 'error';

export interface CaseDoc {
  key?: string;
  filename: string;
  size: number;
  contentType: string;
  uploadedAt: string;
}

export interface EligibilityItem {
  name: string;
  eligible: boolean | null;
  missing_fields: string[];
  estimated_amount?: number;
  reasoning?: string[];
  next_steps?: string;
  requiredForms?: string[];
  eligibility_percent?: number;
  status?: 'eligible' | 'conditional' | 'ineligible';
  rationale?: string;
  certainty_level?: number;
}

export interface CaseSnapshot {
  caseId: string | null;
  status?: CaseStatus;
  documents?: CaseDoc[];
  analyzerFields?: Record<string, unknown>;
  questionnaire?: {
    data?: Record<string, any>;
    missingFieldsHint?: string[];
    lastUpdated?: string;
  };
  eligibility?: EligibilityItem[];
  generatedForms?: Array<{ name: string; status?: string; link?: string }>;
  createdAt?: string;
  updatedAt?: string;
}

export type ResultsEnvelope = {
  results: EligibilityItem[];
  requiredForms: string[];
};

export type EligibilityReport = ResultsEnvelope | EligibilityItem[];

