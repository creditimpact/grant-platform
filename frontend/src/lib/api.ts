import axios from 'axios';

const base = process.env.NEXT_PUBLIC_API_BASE || 'http://localhost:5000';

export const api = axios.create({
  baseURL: `${base.replace(/\/+$/, '')}/api`,
});

