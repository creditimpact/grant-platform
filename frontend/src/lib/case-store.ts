import { create } from 'zustand';

interface CaseState {
  caseId: string | null;
  setCaseId: (id: string) => void;
  clearCaseId: () => void;
}

export const useCaseStore = create<CaseState>((set) => ({
  caseId: null,
  setCaseId: (id: string) => {
    set({ caseId: id });
    if (typeof window !== 'undefined') {
      try {
        localStorage.setItem('caseId', id);
      } catch {}
    }
  },
  clearCaseId: () => {
    set({ caseId: null });
    if (typeof window !== 'undefined') {
      try {
        localStorage.removeItem('caseId');
      } catch {}
    }
  },
}));

export function getCaseId(loadFromStorage = false): string | null {
  const { caseId } = useCaseStore.getState();
  if (caseId) return caseId;
  if (loadFromStorage && typeof window !== 'undefined') {
    try {
      const stored = localStorage.getItem('caseId');
      if (stored) {
        useCaseStore.setState({ caseId: stored });
        return stored;
      }
    } catch {}
  }
  return null;
}

export function setCaseId(id: string) {
  useCaseStore.getState().setCaseId(id);
}

export function clearCaseId() {
  useCaseStore.getState().clearCaseId();
}


