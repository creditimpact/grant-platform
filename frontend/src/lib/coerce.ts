export const toNumberOrUndefined = (v: unknown): number | undefined => {
  if (v === null || v === undefined || v === '') return undefined;
  const n = Number(v);
  return Number.isFinite(n) ? n : undefined;
};

export const yesNoToBoolOrUndefined = (v: unknown): boolean | undefined => {
  if (v === null || v === undefined || v === '') return undefined;
  const s = String(v).toLowerCase();
  if (s === 'yes' || s === 'true') return true;
  if (s === 'no' || s === 'false') return false;
  return undefined;
};

export const keepStringOrEmpty = (v: unknown): string =>
  v === null || v === undefined ? '' : String(v);
