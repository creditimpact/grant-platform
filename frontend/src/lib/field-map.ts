// Maps API (snake_case) → UI (camelCase) and vice versa.
export const apiToUi: Record<string, string> = {
  business_type: 'businessType',
  number_of_employees: 'numberOfEmployees',
  ownership_percentage: 'ownershipPercentage',
  incorporation_date: 'incorporationDate',
  service_area_population: 'serviceAreaPopulation',
  location_zone: 'locationZone',
  project_type: 'projectType',
  project_cost: 'projectCost',
  project_state: 'projectState',
  annual_revenue: 'annualRevenue',
  net_profit: 'netProfit',
  business_income: 'businessIncome',
  business_expenses: 'businessExpenses',
  tax_paid: 'taxPaid',
  tax_year: 'taxYear',
  previous_grants: 'previousGrants',
  previous_refunds_claimed: 'previousRefundsClaimed',
  ein: 'ein',
  ssn: 'ssn',
  sam: 'sam',
  cage_code: 'cageCode',
  duns: 'duns',
  state: 'state',
  zip: 'zip',
  // add more as you see in API payloads
};

export const uiToApi: Record<string, string> = Object.fromEntries(
  Object.entries(apiToUi).map(([k, v]) => [v, k])
);

// Convert snake_case to camelCase if missing above
export const snakeToCamel = (s: string) =>
  s.replace(/_([a-z])/g, (_, c) => c.toUpperCase());

// Normalize options from analyzer to UI selects (e.g., 'llc' → 'LLC')
export const normalizeBusinessTypeForUI = (val?: string) => {
  if (!val) return '';
  const v = val.trim().toLowerCase();
  if (['llc'].includes(v)) return 'LLC';
  if (['sole', 'sole-prop', 'sole proprietorship'].includes(v)) return 'Sole';
  if (['partnership'].includes(v)) return 'Partnership';
  if (['corp', 'corporation', 'c-corp', 's-corp', 's corporation'].includes(v)) return 'Corporation';
  return '';
};

// Normalize dates to YYYY-MM-DD if they look like dd/MM/yyyy or MM/dd/yyyy
export const toIsoDateYMD = (val?: string): string | undefined => {
  if (!val) return undefined;
  const v = val.trim();
  if (/^\d{4}-\d{2}-\d{2}$/.test(v)) return v;
  const slash = v.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})$/);
  if (slash) {
    const [_, a, b, yyyy] = slash;
    // Best-effort: assume dd/MM or MM/dd based on >12 day/month heuristics
    const d = parseInt(a, 10);
    const m = parseInt(b, 10);
    const dayFirst = d > 12 || (d <= 12 && m <= 12 ? false : true);
    const day = String(dayFirst ? d : m).padStart(2, '0');
    const month = String(dayFirst ? m : d).padStart(2, '0');
    return `${yyyy}-${month}-${day}`;
  }
  return v; // leave as-is
};
