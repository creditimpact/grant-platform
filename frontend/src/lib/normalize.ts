export function toArray(value: string | string[] | null | undefined): string[] {
  if (value == null) return [];
  return Array.isArray(value) ? value.filter(Boolean) : [value].filter(Boolean);
}

export function normalizeEligibility<T extends { missing_fields?: string | string[] }>(
  arr: T[]
): (Omit<T, 'missing_fields'> & { missing_fields: string[] })[] {
  return (arr ?? []).map((item) => ({
    ...item,
    missing_fields: toArray(item.missing_fields),
  }));
}
