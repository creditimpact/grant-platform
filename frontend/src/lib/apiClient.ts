import axios from 'axios';
import { normalizeEligibility } from './normalize';
import type {
  CaseSnapshot,
  EligibilityReport,
  GeneratedForm,
  ResultsEnvelope,
} from './types';

const base = process.env.NEXT_PUBLIC_API_BASE || 'http://localhost:5000';

export const api = axios.create({
  baseURL: `${base.replace(/\/+$/, '')}/api`,
});

export async function getStatus(caseId?: string | null): Promise<CaseSnapshot> {
  if (!caseId) {
    return { caseId: null, status: 'empty' };
  }
  const url = `/case/status?caseId=${caseId}`;
  const res = await api.get(url);
  return transformCase(res.data);
}

export async function initCase(): Promise<CaseSnapshot> {
  const res = await api.post('/case/init');
  return transformCase(res.data);
}

export async function getRequiredDocuments(caseId: string): Promise<string[]> {
  const res = await api.get(`/case/required-documents?caseId=${caseId}`);
  return res.data.required;
}

// -------------------- FILE UPLOAD --------------------
export async function uploadFile(formData: FormData): Promise<CaseSnapshot> {
  const res = await api.post('/files/upload', formData);
  return transformCase(res.data);
}

// -------------------- QUESTIONNAIRE --------------------
export async function postQuestionnaire(payload: {
  caseId?: string;
  answers: Record<string, any>;
}): Promise<CaseSnapshot> {
  const res = await api.post('/questionnaire', payload);
  return transformCase(res.data);
}

// -------------------- ELIGIBILITY REPORT --------------------
export async function postEligibilityReport(payload: {
  caseId: string;
}): Promise<CaseSnapshot> {
  const res = await api.post('/eligibility-report', payload);
  return transformCase(res.data);
}

export async function getEligibilityReport(caseId?: string): Promise<EligibilityReport> {
  const r = await fetch(
    `/api/eligibility-report${caseId ? `?caseId=${caseId}` : ''}`,
    { cache: 'no-store' }
  );
  if (!r.ok) throw new Error(`eligibility-report ${r.status}`);

  const data = await r.json();

  // Legacy: API returned raw array
  if (Array.isArray(data)) {
    return {
      results: normalizeEligibility(data),
      requiredForms: [],
    } as EligibilityReport;
  }

  // Envelope style
  const envelope = data as ResultsEnvelope;
  if (Array.isArray(envelope.results)) {
    envelope.results = normalizeEligibility(envelope.results);
  } else {
    envelope.results = [];
  }
  if (!Array.isArray(envelope.requiredForms)) {
    envelope.requiredForms = [];
  }
  return envelope as EligibilityReport;
}

// -------------------- FORM FILL --------------------
export interface PostFormFillResponse {
  generatedForms: GeneratedForm[];
}

export async function postFormFill(
  caseId: string,
  forms: string[],
): Promise<PostFormFillResponse> {
  const res = await fetch('/api/case/form-fill', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ caseId, forms }),
    credentials: 'include',
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err?.message ?? `form-fill failed (${res.status})`);
  }
  const data = await res.json();
  return { generatedForms: toGeneratedForms(data.generatedForms) };
}

// -------------------- TRANSFORM CASE --------------------
function toGeneratedForms(arr: any): GeneratedForm[] {
  return (Array.isArray(arr) ? arr : []).map((f: any) => ({
    formId: f.formId ?? f.formKey ?? '',
    name: f.name ?? f.formId ?? f.formKey ?? '',
    url: f.url ?? f.link ?? '',
    version: f.version ?? f.formVersion ?? undefined,
  }));
}

function transformCase(data: any): CaseSnapshot {
  const rawResults =
    Array.isArray(data.eligibility?.results)
      ? data.eligibility.results
      : Array.isArray(data.eligibility)
      ? data.eligibility
      : [];

  const questionnaire = {
    data: data.questionnaire?.data || {},
    missingFieldsHint: data.questionnaire?.missingFieldsHint || [],
    lastUpdated: data.questionnaire?.lastUpdated,
  };

  return {
    caseId: data.caseId,
    requiredForms: Array.isArray(data.requiredForms) ? data.requiredForms : undefined,
    status: data.status,
    documents: data.documents || [],
    analyzerFields: data.analyzerFields || data.analyzer?.fields || {},
    questionnaire,
    eligibility: rawResults.length ? normalizeEligibility(rawResults) : [],
    generatedForms: toGeneratedForms(data.generatedForms),
    createdAt: data.createdAt,
    updatedAt: data.updatedAt,
  };
}
