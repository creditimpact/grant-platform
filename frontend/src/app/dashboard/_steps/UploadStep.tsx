'use client';
import { useEffect, useState } from 'react';
import { uploadFile, getRequiredDocuments } from '@/lib/apiClient';
import type { CaseDoc, CaseSnapshot } from '@/lib/types';

export default function UploadStep({
  caseId,
  docs,
  onUploaded,
  onNext,
  onBack,
}: {
  caseId: string;
  docs: CaseDoc[];
  onUploaded: (snap: CaseSnapshot) => void;
  onNext: () => void;
  onBack: () => void;
}) {
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | undefined>();
  const [required, setRequired] = useState<string[]>([]);

  useEffect(() => {
    getRequiredDocuments(caseId).then(setRequired).catch(() => setRequired([]));
  }, [caseId]);

  const handleUpload = async (
    reqDoc: string,
    e: React.ChangeEvent<HTMLInputElement>
  ) => {
    const file = e.target.files?.[0];
    if (!file) return;
    const fd = new FormData();
    fd.append('file', file);
    fd.append('caseId', caseId);
    fd.append('key', reqDoc);
    setLoading(true);
    setError(undefined);
    try {
      const snap = await uploadFile(fd);
      onUploaded(snap);
    } catch (err: any) {
      setError(err?.response?.data?.message || err.message || 'Upload failed');
    } finally {
      setLoading(false);
      e.target.value = '';
    }
  };

  const allUploaded =
    required.length > 0 &&
    required.every((r) => docs.some((d) => d.key === r));

  return (
    <div className="space-y-4">
      <h2 className="text-2xl font-bold">Upload Documents</h2>
      {error && <div className="bg-red-100 text-red-800 p-2 rounded">{error}</div>}
      <ul className="space-y-2">
        {required.map((r) => {
          const uploaded = docs.some((d) => d.key === r);
          const slug = r.replace(/[^a-z0-9]/gi, '_');
          return (
            <li key={r} className="flex items-center space-x-2 text-sm">
              <span>{uploaded ? '✅' : '⬜'} {r}</span>
              {!uploaded && (
                <input
                  type="file"
                  data-testid={`upload-${slug}`}
                  onChange={(e) => handleUpload(r, e)}
                  disabled={loading}
                />
              )}
            </li>
          );
        })}
      </ul>
      <div className="flex justify-between">
        <button
          type="button"
          onClick={onBack}
          className="px-3 py-2 rounded bg-gray-200"
        >
          Back
        </button>
        <button
          type="button"
          onClick={onNext}
          disabled={!allUploaded}
          className="px-3 py-2 rounded bg-blue-600 text-white disabled:opacity-50"
        >
          Next
        </button>
      </div>
    </div>
  );
}
