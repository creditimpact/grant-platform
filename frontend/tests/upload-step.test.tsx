import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import UploadStep from '@/app/dashboard/_steps/UploadStep';
import * as api from '@/lib/apiClient';
import type { CaseDoc } from '@/lib/types';

jest.mock('@/lib/apiClient');

describe('UploadStep', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders required documents', async () => {
    (api.getRequiredDocuments as jest.Mock).mockResolvedValue([
      'Tax Returns',
      'Payroll Records',
    ]);
    const { asFragment } = render(
      <UploadStep
        caseId="c1"
        docs={[]}
        onUploaded={jest.fn()}
        onNext={jest.fn()}
        onBack={jest.fn()}
      />
    );
    await screen.findByText(/Tax Returns/);
    await screen.findByText(/Payroll Records/);
    expect(asFragment()).toMatchSnapshot();
  });

  it('updates checklist after uploads', async () => {
    (api.getRequiredDocuments as jest.Mock).mockResolvedValue([
      'Tax Returns',
      'Payroll Records',
    ]);
    const docs: CaseDoc[] = [];
    (api.uploadFile as jest.Mock).mockImplementation(async (fd: FormData) => {
      const key = fd.get('key') as string;
      const file = fd.get('file') as File;
      docs.push({
        key,
        filename: file.name,
        size: file.size,
        contentType: file.type,
        uploadedAt: 'now',
      });
      return { caseId: 'c1', documents: docs };
    });
    let rerenderFn: any;
    const handleUploaded = (snap: any) => {
      docs.splice(0, docs.length, ...snap.documents);
      rerenderFn(
        <UploadStep
          caseId="c1"
          docs={docs}
          onUploaded={handleUploaded}
          onNext={jest.fn()}
          onBack={jest.fn()}
        />
      );
    };
    const { rerender } = render(
      <UploadStep
        caseId="c1"
        docs={docs}
        onUploaded={handleUploaded}
        onNext={jest.fn()}
        onBack={jest.fn()}
      />
    );
    rerenderFn = rerender;
    const taxInput = await screen.findByTestId('upload-Tax_Returns');
    const file1 = new File(['a'], 'tax.pdf', { type: 'application/pdf' });
    fireEvent.change(taxInput, { target: { files: [file1] } });
    await waitFor(() => expect(api.uploadFile).toHaveBeenCalled());
    expect(screen.getByText(/Tax Returns/).textContent).toContain('✅');
    const payrollInput = screen.getByTestId('upload-Payroll_Records');
    const file2 = new File(['b'], 'payroll.pdf', { type: 'application/pdf' });
    fireEvent.change(payrollInput, { target: { files: [file2] } });
    await waitFor(() => expect(api.uploadFile).toHaveBeenCalledTimes(2));
    expect(screen.getByText(/Payroll Records/).textContent).toContain('✅');
    const nextBtn = screen.getByText('Next') as HTMLButtonElement;
    await waitFor(() => expect(nextBtn.disabled).toBe(false));
  });
});
