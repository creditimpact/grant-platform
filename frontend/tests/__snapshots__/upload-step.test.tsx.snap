// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadStep renders required documents 1`] = `
<DocumentFragment>
  <div
    class="space-y-4"
  >
    <h2
      class="text-2xl font-bold"
    >
      Upload Documents
    </h2>
    <ul
      class="space-y-2"
    >
      <li
        class="flex items-center space-x-2 text-sm"
      >
        <span>
          ⬜ Tax Returns
        </span>
        <input
          data-testid="upload-Tax_Returns"
          type="file"
        />
      </li>
      <li
        class="flex items-center space-x-2 text-sm"
      >
        <span>
          ⬜ Payroll Records
        </span>
        <input
          data-testid="upload-Payroll_Records"
          type="file"
        />
      </li>
    </ul>
    <div
      class="flex justify-between"
    >
      <button
        class="px-3 py-2 rounded bg-gray-200"
        type="button"
      >
        Back
      </button>
      <button
        class="px-3 py-2 rounded bg-blue-600 text-white disabled:opacity-50"
        disabled=""
        type="button"
      >
        Next
      </button>
    </div>
  </div>
</DocumentFragment>
`;
